src/shape_transform_descriptor.cpp

```diff
@@ -0,0 +1,836 @@
+                subs[i]->axis = {axis, i};
+        }
+    }
+}
+static void renumber_axes(std::vector<dimension>& dimensions)
+{
+    auto axes_map = group_axes(dimensions);
+    renumber_axes(axes_map);
+}
+static void collapse_1_dims(std::vector<dimension>& dimensions)
+{
+    // Find a dimension that ends with a subdimension of 1 with a single axis,
+    // and is followed by subdimension in the next dimension of 1 that has a
+    // split axis. It will remove the trailing subdimension and update the
+    // leading subdimension to use the axis from the trailing subdimension.
+    adjacent_for_each(dimensions.begin(), dimensions.end(), [&](dimension& d1, dimension& d2) {
+        if(d1.subdimensions.size() < 2)
+            return;
+        if(d2.subdimensions.empty())
+            return;
+        if(d2.len() != 1)
+            return;
+        const auto& sub1 = d1.subdimensions.back();
+        auto& sub2       = d2.subdimensions.front();
+        if(sub1.axis.size() != 1)
+            return;
+        if(sub2.axis.size() < 2)
+            return;
+        if(sub1.len != 1)
+            return;
+        if(sub2.len != 1)
+            return;
+        sub2.axis = sub1.axis;
+        d1.subdimensions.pop_back();
+    });
+
+    renumber_axes(dimensions);
+}
+
+void shape_transform_descriptor::simplify()
+{
+    for(auto& d : dimensions)
+        d.simplify();
+
+    std::map<std::size_t, std::size_t> missing_axes;
+    std::vector<std::size_t> last_axis;
+    {
+        // Group axes
+        auto axes_map = group_axes(dimensions);
+        if(axes_map.empty())
+            return;
+
+        renumber_axes(axes_map);
+
+        // Find last axis
+        last_axis = std::prev(axes_map.end())->second.back()->axis;
+
+        // Find missing axes. This will store a mapping between the missing
+        // axis and the next available axis.
+        for(auto axis : range(rank))
+        {
+            if(contains(axes_map, axis))
+                continue;
+            auto it            = axes_map.upper_bound(axis);
+            missing_axes[axis] = it == axes_map.end() ? rank : it->first;
+        }
+    }
+
+    // Find broadcasted dimensions. This will store a map from the next axis
+    // to the previous axes being broadcasted.
+    std::map<std::size_t, std::deque<std::size_t>> broadcast_dims_map;
+    group_find(
+        dimensions.begin(), dimensions.end(), &missing_leading_axis, [&](auto start, auto last) {
+            auto axis = rank;
+            if(last != dimensions.end())
+            {
+                assert(not last->subdimensions.empty());
+                const auto& sub = last->subdimensions.front();
+                assert(not sub.axis.empty());
+                axis = sub.axis.front();
+            }
+            std::deque<std::size_t> dims(std::distance(start, last));
+            std::iota(dims.begin(), dims.end(), std::distance(dimensions.begin(), start));
+            broadcast_dims_map[axis] = dims;
```

Comment:
Would this work in case of transposes ? When there are transposes, dimensions would not be in sorted order based on original axis.

Reply:
This is meant to store the index of the dimension. Let me update the comment to make that clearer.